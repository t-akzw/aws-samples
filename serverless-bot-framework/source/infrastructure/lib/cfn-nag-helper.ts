/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import { CfnResource } from '@aws-cdk/core';

export interface CfnNagSuppression {
  Id: string;
  Reason: string;
}

export class CfnNagHelper {
  public static addSuppressions(resource: CfnResource, suppressions: CfnNagSuppression | CfnNagSuppression[]) {

    let rules = [];

    if (suppressions instanceof Array) {
      for (const suppression of suppressions) {
        rules.push({ id: suppression.Id, reason: suppression.Reason });
      }
    } else {
      rules.push({ id: suppressions.Id, reason: suppressions.Reason });
    }

    const existingRules = resource.cfnOptions.metadata?.cfn_nag?.rules_to_suppress;
    resource.cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: existingRules ? [...existingRules, ...rules] : rules
      }
    };
  }
}