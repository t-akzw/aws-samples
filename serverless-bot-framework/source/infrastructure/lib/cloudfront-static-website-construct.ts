/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Construct } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';

export class CloudfrontStaticWebsite extends Construct {
  private readonly _webclient: Bucket | undefined;
  private readonly _domainName: string;
  private readonly _bucketName: any;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const cloudfrontToS3 = new CloudFrontToS3(this, 'CloudFrontToS3', {
      insertHttpSecurityHeaders: false,
    });

    this._webclient = cloudfrontToS3.s3Bucket;
    this._domainName = `https://${cloudfrontToS3.cloudFrontWebDistribution.domainName}`;
    this._bucketName = cloudfrontToS3.s3Bucket?.bucketName;
  }

  public get webclient(): Bucket | undefined {
    return this._webclient;
  }

  public get domainName(): string {
    return this._domainName;
  }

  public get bucketName(): string {
    return this._bucketName.toString();
  }
}
