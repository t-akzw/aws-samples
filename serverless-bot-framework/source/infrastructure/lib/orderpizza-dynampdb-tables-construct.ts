/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Construct, RemovalPolicy } from '@aws-cdk/core';
import { Function } from '@aws-cdk/aws-lambda';
import {
  TableProps,
  Table,
  GlobalSecondaryIndexProps,
  AttributeType,
  BillingMode,
  ProjectionType,
} from '@aws-cdk/aws-dynamodb';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';

export interface OrderPizzaLambdaDynamoDBTablesProps {
  readonly orderPizzaLambda: Function;
}

export class OrderPizzaLambdaDynamoDBTables extends Construct {
  private readonly _orderPizzaLambda: Function;
  private readonly _ordersDBTable: Table;

  constructor(
    scope: Construct,
    id: string,
    props: OrderPizzaLambdaDynamoDBTablesProps
  ) {
    super(scope, id);

    /** pass in orderPizzaLambda Function */
    this._orderPizzaLambda = props.orderPizzaLambda;


    /** OrderPizzaLambda Orders DynamoDBTable integration */
    const ordersDBTableProbs: TableProps = {
      partitionKey: {
        name: 'orderId',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
    };

    const ordersDBTableGSI: GlobalSecondaryIndexProps = {
      indexName: 'customerId-orderTimestamp-index',
      partitionKey: {
        name: 'customerId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'orderTimestamp',
        type: AttributeType.NUMBER,
      },
      projectionType: ProjectionType.ALL,
    };

    const orderPizzaOrdersDBTable = new LambdaToDynamoDB(
      this,
      'OrderPizzaOrdersDBTable',
      {
        existingLambdaObj: this._orderPizzaLambda,
        dynamoTableProps: ordersDBTableProbs,
      }
    );

    orderPizzaOrdersDBTable.dynamoTable.addGlobalSecondaryIndex(
      ordersDBTableGSI
    );


    /** Get dynamoDB Tables */
    this._ordersDBTable = orderPizzaOrdersDBTable.dynamoTable;
    // this._menusDBTable = orderPizzaMenusDBTable.dynamoTable;

    /** Add DynamoDB Table and Index names to OrderPizza Lambda's enviroment */
    this._orderPizzaLambda.addEnvironment(
      'PIZZA_ORDERS_TABLE',
      this._ordersDBTable.tableName
    );
    this._orderPizzaLambda.addEnvironment(
      'PIZZA_ORDERS_INDEX',
      ordersDBTableGSI.indexName
    );

  }

  public get orderPizzaLambda(): Function {
    return this._orderPizzaLambda;
  }

  public get ordersDBTable(): Table {
    return this._ordersDBTable;
  }

}
