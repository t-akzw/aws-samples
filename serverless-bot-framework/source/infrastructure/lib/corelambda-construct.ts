/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Aws, Construct } from '@aws-cdk/core';
import { Function, FunctionProps, CfnFunction } from '@aws-cdk/aws-lambda';
import { PolicyStatement, Effect, Policy } from '@aws-cdk/aws-iam';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import { CfnNagHelper } from './cfn-nag-helper';


export interface CoreLambdaProps {
  readonly existingLambdaObj?: Function;
  readonly lambdaFunctionProps?: FunctionProps;
}

export class CoreLambda extends Construct {
  private readonly _coreLambda: Function;

  constructor(scope: Construct, id: string, props: CoreLambdaProps) {
    super(scope, id);

    const coreLambda = buildLambdaFunction(
      this,
      {
        existingLambdaObj: props.existingLambdaObj,
        lambdaFunctionProps: props.lambdaFunctionProps,
      }
    );

    const lexPolicy = new Policy(this, 'LexPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['lex:RecognizeText'],
          resources: [`arn:${Aws.PARTITION}:lex:${Aws.REGION}:${Aws.ACCOUNT_ID}:*`],
        }),
      ],
    });

    /** Grant CoreLambda Lex permissions */
    coreLambda.role?.attachInlinePolicy(lexPolicy); //NOSONAR it is a valid expression
    /** Get the CoreLambda created by the LambdaToS3 construct */
    this._coreLambda = coreLambda;
    /** Suppression for cfn nag */
    const cfnFunction = this._coreLambda.node.defaultChild as CfnFunction;
    CfnNagHelper.addSuppressions(cfnFunction, [
      {
        Id: 'W92',
        Reason: 'This function does not need to have specified reserved concurrent executions'
      }
    ]);

  }

  public get coreLambda(): Function {
    return this._coreLambda;
  }

}
