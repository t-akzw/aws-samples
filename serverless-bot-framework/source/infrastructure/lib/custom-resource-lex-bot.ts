/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Aws, Construct, Duration, CustomResource } from '@aws-cdk/core';
import { Runtime, Code, CfnFunction } from '@aws-cdk/aws-lambda';
import { PolicyStatement, Effect, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import { CfnNagHelper } from './cfn-nag-helper';

export interface LexBotProps {
    readonly botName: string;
    readonly botLanguage: string;
    readonly childDirected: string;
    readonly lexLambdaARN: string;
}

export class LexCustomResource extends Construct {
  private CustomResource: CustomResource;

  public get BotId(): string {
    return this.CustomResource.getAttString('BotId');
  }

  public get BotAliasId(): string {
    return this.CustomResource.getAttString('BotAliasId');
  }

  constructor(scope: Construct, id: string, props: LexBotProps) {
    super(scope, id);

    const LexV2Role = new Role(this, 'LexV2Role', {
      assumedBy: new ServicePrincipal('lexv2.amazonaws.com'),
    });

    const helperFunction = buildLambdaFunction(this, {
      lambdaFunctionProps: {
        runtime: Runtime.PYTHON_3_8,
        handler: 'lambda_function.handler',
        description:
          'This function creates a Lex bot using AWS SDK (boto3).',
        code: Code.fromAsset('../services/lex-bot'),
        timeout: Duration.seconds(60),
        environment: {
          botName: props.botName,
          botLanguage: props.botLanguage,
          childDirected: props.childDirected,
          botRole: LexV2Role.roleArn,
          lexLambdaARN: props.lexLambdaARN,
        }
      },
    });

    /** Grant permission to the lambda function to create Lex Bot */
    helperFunction.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: [
          `arn:${Aws.PARTITION}:lex:${Aws.REGION}:${Aws.ACCOUNT_ID}:bot/*`,
          `arn:${Aws.PARTITION}:lex:${Aws.REGION}:${Aws.ACCOUNT_ID}:bot-alias/*`,
          LexV2Role.roleArn,
        ],
        actions: [
          'lex:CreateBot', 'lex:CreateBotVersion', 'lex:CreateBotLocale', 'lex:CreateSlotType',
          'lex:CreateSlot', 'lex:CreateIntent',
          'lex:DeleteBot', 'lex:DeleteBotLocale', 'lex:DeleteIntent', 'lex:DeleteSlot',
          'lex:DeleteBotVersion', 'lex:DeleteBotChannel', 'lex:DeleteSlotType',
          'lex:DescribeBot', 'lex:DescribeBotLocale',
          'lex:UpdateIntent', 'lex:UpdateBotAlias',
          'lex:ListBotAliases',
          'lex:TagResource',
          'lex:BuildBotLocale',
          'iam:PassRole'
        ],
      })
    );
    helperFunction.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: [
          /** ListBots requires service level permission for lex hence the
           * lex:region:accoun:* specified for resource. It is separated from
           * the rest of Lex actions to prevent service level permission for
           * other Lex actions.
           */
          `arn:${Aws.PARTITION}:lex:${Aws.REGION}:${Aws.ACCOUNT_ID}:*`,
        ],
        actions:['lex:ListBots']
      })
    );

    /** Grant permission to the lex bot to invoke feedback lambda function */
    LexV2Role.addToPolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: [props.lexLambdaARN],
        actions: ['lambda:InvokeFunction']
      })
    )

    this.CustomResource = new CustomResource(this, 'LexCustomResource', {
        serviceToken: helperFunction.functionArn,
        resourceType: 'Custom::LexBotCustomResource',
        properties: {
          botLanguage: props.botLanguage,
          botName: props.botName,
          childDirected: props.childDirected,
        }
    });

    /** Suppression for cfn nag W92 */
    const cfnFunction = helperFunction.node.defaultChild as CfnFunction;
    CfnNagHelper.addSuppressions(cfnFunction, {
        Id: 'W92',
        Reason: 'This function does not need to have specified reserved concurrent executions'
    });
  }

}
