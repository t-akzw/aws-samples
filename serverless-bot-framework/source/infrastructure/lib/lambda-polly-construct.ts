/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Construct } from '@aws-cdk/core';
import { Function, FunctionProps, CfnFunction } from '@aws-cdk/aws-lambda';
import { PolicyStatement, Effect, Policy, CfnPolicy } from '@aws-cdk/aws-iam';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import { CfnNagHelper } from './cfn-nag-helper';

export interface LambdaToPollyProps {
  readonly existingLambdaObj?: Function;
  readonly lambdaFunctionProps?: FunctionProps;
}

export class LambdaToPolly extends Construct {
  private readonly _pollyLambda: Function;

  constructor(scope: Construct, id: string, props: LambdaToPollyProps) {
    super(scope, id);

    this._pollyLambda = buildLambdaFunction(this, {
      existingLambdaObj: props.existingLambdaObj,
      lambdaFunctionProps: props.lambdaFunctionProps,
    });

    const pollyPolicy = new Policy(this, 'PollyPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['polly:SynthesizeSpeech'],
          resources: ['*'],
        }),
      ],
    });

    /** Suppression for cfn nag */
    const cfnFunction = this._pollyLambda.node.defaultChild as CfnFunction;
    CfnNagHelper.addSuppressions(cfnFunction, {
      Id: 'W92',
      Reason: 'This function does not need to have specified reserved concurrent executions'
    });

    /** Add policy metadata to explain why resources: ['*'] is needed */
    const cfnPolicy = pollyPolicy.node.defaultChild as CfnPolicy;
    CfnNagHelper.addSuppressions(cfnPolicy, {
      Id: 'W12',
      Reason:
        'Polly allows specifying lexicon ARNs only. There is no specific lexicon required for this policy',
    });

    /** Add the pollyPolicy to the pollyLambda's role */
    this._pollyLambda.role?.attachInlinePolicy(pollyPolicy); //NOSONAR it is a valid expression
  }

  public get pollyLambda(): Function {
    return this._pollyLambda;
  }
}
