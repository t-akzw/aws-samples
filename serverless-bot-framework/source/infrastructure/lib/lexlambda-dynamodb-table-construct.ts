/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Construct, Aws, RemovalPolicy } from '@aws-cdk/core';
import { Function, FunctionProps, CfnFunction } from '@aws-cdk/aws-lambda';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import {
  TableProps,
  Table,
  AttributeType,
  BillingMode,
} from '@aws-cdk/aws-dynamodb';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import { CfnNagHelper } from './cfn-nag-helper';


export interface LexLambdaDynamoDBTableProps {
  readonly lexLambdaProps: FunctionProps;
}

export class LexLambdaDynamoDBTable extends Construct {
  private readonly _lexLambda: Function;
  private readonly _lexDBTable: Table;

  constructor(
    scope: Construct,
    id: string,
    props: LexLambdaDynamoDBTableProps
  ) {
    super(scope, id);

    /** build lexLambda Function */
    this._lexLambda = buildLambdaFunction(this, {
      lambdaFunctionProps: props.lexLambdaProps,
    });

    /** add resource policy to allow lexv2 bot invoke this function */
    this._lexLambda.addPermission(
      'lexv2permission',
      {
        action: 'lambda:InvokeFunction',
        principal: new ServicePrincipal('lexv2.amazonaws.com'),
        sourceArn: `arn:${Aws.PARTITION}:lex:${Aws.REGION}:${Aws.ACCOUNT_ID}:bot-alias/*`,
      });

    /** lexLambda lex DynamoDBTable integration */
    const lexDBTableProbs: TableProps = {
      partitionKey: {
        name: 'uuid',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
    };

    const lexLambdaDBTable = new LambdaToDynamoDB(
      this,
      'LexLambdaDBTable',
      {
        existingLambdaObj: this._lexLambda,
        dynamoTableProps: lexDBTableProbs,
      }
    );

    /** Get dynamoDB Table */
    this._lexDBTable = lexLambdaDBTable.dynamoTable;
    this._lexLambda.addEnvironment('FEEDBACK_TABLE',
      this._lexDBTable.tableName
    );

    /** Suppression for cfn nag W92 */
    const cfnFunction = this._lexLambda.node.defaultChild as CfnFunction;
    CfnNagHelper.addSuppressions(cfnFunction, {
        Id: 'W92',
        Reason: 'This function does not need to have specified reserved concurrent executions'
    });
  }

  public get lexLambda(): Function {
    return this._lexLambda;
  }

  public get lexDBTable(): Table {
    return this._lexDBTable;
  }
}
