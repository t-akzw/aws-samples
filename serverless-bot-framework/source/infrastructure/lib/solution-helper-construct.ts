/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import {
  Aws,
  Construct,
  CustomResource,
  CfnCondition,
  Duration,
} from '@aws-cdk/core';
import { Runtime, Code, CfnFunction } from '@aws-cdk/aws-lambda';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import { CfnNagHelper } from './cfn-nag-helper';

export interface SolutionHelperProps {
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly sendAnonymousDataCondition: CfnCondition;
  readonly botName: string;
  readonly botGender: string;
  readonly botLanguage: string;
}

export class SolutionHelper extends Construct {
  private readonly _createIdFunction: CustomResource;

  constructor(scope: Construct, id: string, props: SolutionHelperProps) {
    super(scope, id);

    const helperFunction = buildLambdaFunction(this, {
      lambdaFunctionProps: {
        runtime: Runtime.PYTHON_3_8,
        handler: 'lambda_function.handler',
        description:
          'This function generates UUID for each deployment and sends anonymous data to the AWS Solutions team',
        code: Code.fromAsset('../services/solution-helper'),
        timeout: Duration.seconds(30),
      },
    });

    this._createIdFunction = new CustomResource(this, 'CreateUniqueID', {
      serviceToken: helperFunction.functionArn,
      properties: {
        Resource: 'UUID',
      },
      resourceType: 'Custom::CreateUUID',
    });

    const sendDataFunction = new CustomResource(this, 'SendAnonymousData', {
      serviceToken: helperFunction.functionArn,
      properties: {
        Resource: 'AnonymousMetric',
        SolutionId: props.solutionId,
        UUID: this._createIdFunction.getAttString('UUID'),
        Region: Aws.REGION,
        Version: props.solutionVersion,
        botName: props.botName,
        botGender: props.botGender,
        botLanguage: props.botLanguage,
      },
      resourceType: 'Custom::AnonymousData',
    });

    /** Suppression for cfn nag W92 */
    const cfnFunction = helperFunction.node.defaultChild as CfnFunction;
    CfnNagHelper.addSuppressions(cfnFunction, {
        Id: 'W92',
        Reason: 'This function does not need to have specified reserved concurrent executions'
    });
    /** Add the metricsCondition to the resources */
    (helperFunction.node.defaultChild as CfnFunction).cfnOptions.condition =
      props.sendAnonymousDataCondition;
    (this._createIdFunction.node
      .defaultChild as CfnFunction).cfnOptions.condition =
      props.sendAnonymousDataCondition;
    (sendDataFunction.node.defaultChild as CfnFunction).cfnOptions.condition =
      props.sendAnonymousDataCondition;

  }

  public get createIdFunction(): CustomResource {
    return this._createIdFunction;
  }
}
