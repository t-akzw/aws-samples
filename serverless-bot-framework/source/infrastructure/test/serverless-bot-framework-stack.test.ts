/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { ServerlessBotFrameworkStack } from '../lib/serverless-bot-framework-stack';
import '@aws-cdk/assert/jest';
import { execSync } from "child_process";

beforeAll(() => {
  /** create a fake build directory for the webclient package */
  execSync(`mkdir ${__dirname}/../../samples/webclient/build`);
});

afterAll(() => {
  /** remove the fake build directory for the webclient package */
  execSync(`rm -rf ${__dirname}/../../samples/webclient/build`);
});

test('test ServerlessBotFrameworkStack stack', () => {
  const stack = new Stack();
  new ServerlessBotFrameworkStack(stack, 'ServerlessBotFrameworkStack', {
    solutionID: 'SO0027',
    solutionName: 'UnitTestExecution',
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
