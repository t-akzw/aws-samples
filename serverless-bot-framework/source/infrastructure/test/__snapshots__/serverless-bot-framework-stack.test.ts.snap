// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ServerlessBotFrameworkStack stack 1`] = `Object {}`;
